src/Promise/PromiseInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+namespace GraphQL\Promise;
+
+interface PromiseInterface
+{
+    /**
+     * Waits until the promise completes if possible.
+     *
+     * @return mixed
+     * @throws \LogicException if the promise has no wait function.
+     */
+    public function wait();
+
```

Comment:
As for `wait` method. My point about `wait` as a part of `PromiseInterface` is that in `event-loop` environment you do not need to block ever. In ReactPHP you would expect something like this:

```php
GraphQL::execute(...)->then(...)
```

You don't want to block untill all of our promises are resolved.

Also `wait` is simply not a part of common Promise interface. If you check [ReachPHP promise interface](https://github.com/reactphp/promise/blob/master/src/PromiseInterface.php) or [Promises A+ spec](https://promisesaplus.com/) - they do not mention `wait` or `block`.

Guzzle does, but that's because it works in non-event loop environment. And even their promises docs mention `wait` in section called [Synchronous Wait](https://github.com/guzzle/promises#synchronous-wait). They had to add it, because Guzzle HTTP client is often consumed from sync environment where devs are used to do things synchronously.

Of course you could force all promises to implement blocking `wait`, but for platforms with event-loops it would be an unnecessary limitation.

Also I don't think that [your implementation of `await` for ReachPHP](https://github.com/mcg-web/promise-factory/blob/master/src/Factory/ReactPromiseFactory.php#L86-L110) actually does what it claims to do, as it doesn't seem to block.

Are you sure it works as expected in ReactPHP? As for me it looks like it will work only with resolved promises, but not those which are still pending. Though maybe I am missing something.